File: app/src/main/java/com/thatcakeid/zrytezene/LoginActivity.java
package com.thatcakeid.zrytezene;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.FirebaseAuth;

import com.thatcakeid.zrytezene.databinding.ActivityLoginBinding;

public class LoginActivity extends AppCompatActivity {
    private TextInputEditText email_tie, passw_tie;
    private TextInputLayout email_til;
    private FirebaseAuth auth;

    ActivityLoginBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);

        ExtraMetadata.setWatermarkColors(binding.textWatermarkLogin);

        email_tie = binding.loginEmailTie;
        passw_tie = binding.loginPasswTie;
        email_til = binding.loginEmailTil;

        TextInputLayout passw_til = binding.loginPasswTil;

        MaterialButton button_continue = binding.buttonContinue;

        TextView register_text = binding.registerText;
        TextView forgot_password_text = binding.forgotPasswordText;

        FirebaseApp.initializeApp(getApplicationContext());
        auth = FirebaseAuth.getInstance();

        email_tie.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }

            @Override
            public void afterTextChanged(Editable s) { }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (email_tie.getText().toString().trim().matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+")) {
                    email_til.setError(null);
                } else {
                    email_til.setError("Invalid email!");
                }
            }
        });

        button_continue.setOnClickListener(v -> {
            String email = email_tie.getText().toString().trim();
            String password = passw_tie.getText().toString();

            if (!email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+")) {
                Toast.makeText(LoginActivity.this, "Invalid email!", Toast.LENGTH_LONG).show();
                return;
            }

            if (password.length() == 0) {
                //Snackbar.make(view, "Password can't be empty!", Snackbar.LENGTH_LONG);
                Toast.makeText(LoginActivity.this, "Password can't be empty!", Toast.LENGTH_LONG).show();

            } else {
                auth.signInWithEmailAndPassword(email, password).addOnSuccessListener(authResult -> {
                    if (auth.getCurrentUser().isEmailVerified()) {
                        Toast.makeText(LoginActivity.this, "Logged in", Toast.LENGTH_LONG).show();
                        startActivity(new Intent(getApplicationContext(), HomeActivity.class));
                        finish();

                    } else {
                        AlertDialog.Builder alertDialog = new AlertDialog.Builder(LoginActivity.this);

                        alertDialog.setTitle("Unverified Email");
                        alertDialog.setMessage("Your email isn't verified. Do you want to re-send a new verification link to your email?");

                        alertDialog.setPositiveButton("Yes", (dialog, which) -> auth.getCurrentUser().sendEmailVerification().addOnSuccessListener(aVoid -> {
                            //Snackbar.make(view, "A verification link has been sent to your email. Please check your inbox or spam box.", Snackbar.LENGTH_LONG);
                            Toast.makeText(LoginActivity.this, "A verification link has been sent to your email. Please check your inbox or spam box.", Toast.LENGTH_LONG).show();
                        }).addOnFailureListener(e -> {
                            //Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                            Toast.makeText(LoginActivity.this, e.getMessage(), Toast.LENGTH_LONG).show();
                        }));

                        alertDialog.setNegativeButton("No", (dialog, which) -> auth.signOut());
                        alertDialog.setCancelable(false);
                        alertDialog.create().show();
                    }

                }).addOnFailureListener(e -> {
                    //Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                    Toast.makeText(LoginActivity.this, e.getMessage(), Toast.LENGTH_LONG).show();
                });
            }
        });

        forgot_password_text.setOnClickListener(v -> {
            String email = email_tie.getText().toString().trim();

            if (email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+")) {
                Toast.makeText(LoginActivity.this, "Please enter a valid email!", Toast.LENGTH_LONG).show();
                return;
            }

            auth.sendPasswordResetEmail(email);
            //Snackbar.make(view, "A verification link has been sent to your email. Please check your inbox or spam box.", Snackbar.LENGTH_LONG);
            Toast.makeText(LoginActivity.this, "A verification link has been sent to your email. Please check your inbox or spam box.", Toast.LENGTH_LONG).show();
        });

        register_text.setOnClickListener(v -> {
            Intent i = new Intent(getApplicationContext(), RegisterActivity.class);
            i.putExtra("email", email_tie.getText().toString());
            startActivity(i);
        });
    }
}

File: app/src/main/java/com/thatcakeid/zrytezene/LoginActivity.kt
package com.thatcakeid.zrytezene

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.thatcakeid.zrytezene.ExtraMetadata.setWatermarkColors
import com.thatcakeid.zrytezene.databinding.ActivityLoginBinding

class LoginActivity : AppCompatActivity() {
    private val binding by lazy {
        ActivityLoginBinding.inflate(layoutInflater)
    }

    private val auth by lazy { FirebaseAuth.getInstance() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(binding.root)
        setWatermarkColors(binding.textWatermark, binding.watermarkRoot)

        binding.loginEmailTie.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (binding.loginEmailTie.text.toString().trim()
                        .matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                    binding.loginEmailTil.error = null
                } else {
                    binding.loginEmailTil.error = "Invalid email!"
                }
            }
        })

        binding.buttonContinue.setOnClickListener {
            val email = binding.loginEmailTie.text.toString().trim()
            val password = binding.loginEmailTie.text.toString()

            if (!email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                Toast.makeText(this@LoginActivity, "Invalid email!", Toast.LENGTH_LONG).show()
                return@setOnClickListener
            }

            if (password.isEmpty()) {
                Snackbar.make(binding.root, "Password can't be empty!", Snackbar.LENGTH_LONG).show()
            } else {
                auth.signInWithEmailAndPassword(email, password).addOnSuccessListener {
                        if (auth.currentUser!!.isEmailVerified) {
                            Toast.makeText(this@LoginActivity, "Logged in", Toast.LENGTH_LONG)
                                .show()
                            startActivity(Intent(applicationContext, MainActivity::class.java))
                            finish()

                        } else {
                            val alertDialog = AlertDialog.Builder(this@LoginActivity)

                            alertDialog.setTitle("Unverified Email")
                            alertDialog.setMessage("Your email isn't verified. Do you want to re-send a new verification link to your email?")
                            alertDialog.setPositiveButton("Yes") { _, _ ->
                                auth.currentUser!!
                                    .sendEmailVerification().addOnSuccessListener {
                                        Snackbar.make(
                                            binding.root,
                                            "A verification link has been sent to your email. Please check your inbox or spam box.",
                                            Snackbar.LENGTH_LONG
                                        ).show()
                                    }
                                    .addOnFailureListener { e: Exception ->
                                        Snackbar.make(
                                            binding.root,
                                            e.message!!,
                                            Snackbar.LENGTH_LONG
                                        ).show()
                                    }
                            }

                            alertDialog.setNegativeButton("No") { _, _ -> auth.signOut() }
                            alertDialog.setCancelable(false)
                            alertDialog.create().show()
                        }
                    }.addOnFailureListener { e: Exception ->
                    Snackbar.make(
                        binding.root,
                        e.message!!,
                        Snackbar.LENGTH_LONG
                    ).show()
                }
            }
        }

        binding.forgotPasswordText.setOnClickListener {
            val email = binding.loginEmailTie.text.toString().trim()

            if (email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                Toast.makeText(this@LoginActivity, "Please enter a valid email!", Toast.LENGTH_LONG)
                    .show()
                return@setOnClickListener
            }

            auth.sendPasswordResetEmail(email)
            Snackbar.make(
                binding.root,
                "A verification link has been sent to your email. Please check your inbox or spam box.",
                Snackbar.LENGTH_LONG
            ).show()

        }

        binding.registerText.setOnClickListener {
            startActivity(
                Intent(
                    applicationContext, RegisterActivity::class.java
                ).putExtra("email", binding.loginEmailTie.text.toString())
            )
        }
    }
}